src/test/java/com/faforever/client/io/DownloadServiceTest.java

```diff
@@ -129,4 +133,35 @@ public void testDownloadWithMirrors() throws Exception {
     verify(downloadServiceSpy).downloadFile(new URL("https://mirror1.example.com/foo/bar"), targetFile, progressListener, md5sum);
     verify(downloadServiceSpy).downloadFile(new URL("https://example.com/bar"), targetFile, progressListener, md5sum);
   }
+
+  @Test
+  public void testDownloadFile() throws Exception {
+    URL url = mock(URL.class);
+    URLConnection urlConnection = mock(URLConnection.class);
+
+    byte[] DATA = "some content".getBytes();
+    when(url.openStream()).thenReturn(new ByteArrayInputStream(DATA));
+    when(urlConnection.getContentLength()).thenReturn(DATA.length);
+    when(url.openConnection()).thenReturn(urlConnection);
+
+    Path temp = Files.createTempFile("download", ".dat");
+    instance.downloadFile(url, temp, (a, b) -> {}, "00000000000000000000000000000000");
+
+    byte[] data = Files.readAllBytes(temp);
+    assertEquals(data, DATA);
+  }
+
+  @Test(expected = ChecksumMismatchException.class)
+  public void testDownloadFileBadChecksum() throws Exception {
+    URL url = mock(URL.class);
+    URLConnection urlConnection = mock(URLConnection.class);
+
+    byte[] DATA = "some content".getBytes();
+    when(url.openStream()).thenReturn(new ByteArrayInputStream(DATA));
+    when(urlConnection.getContentLength()).thenReturn(DATA.length);
+    when(url.openConnection()).thenReturn(urlConnection);
+
+    Path temp = Files.createTempFile("download", ".dat");
+    instance.downloadFile(url, temp, (a, b) -> {}, "00000000000000000000000000000000");
+  }
```

Comment:
It is better to wrap the instance.downloadFile inside of an Assertions.assertThrows() so that we are checking the specific line we expect to throw the exception rather than the whole method

Reply:
Yea I wasn’t sure how to do it in Java. I’ll update it